### PR TITLE
util: Remove DirIsWritable, GetUniquePath

### PR DESCRIPTION
**Pull request by maflcko**
`GetUniquePath` is only used in tests and in `DirIsWritable`. The check by `DirIsWritable` is redundant with the check done in `LockDirectory`.

Fix the redundancy by removing everything, except `LockDirectory`.

### CODE REVIEW

**Comment by fanquake** [comment]
```bash
make[2]: Entering directory '/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-arm-linux-gnueabihf/src'
/usr/bin/ccache arm-linux-gnueabihf-g++ -std=c++17 -DHAVE_CONFIG_H -I. -I../src/config  -fmacro-prefix-map=/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-arm-linux-gnueabihf=. -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -DHAVE_BUILD_INFO -D_FILE_OFFSET_BITS=64 -DPROVIDE_FUZZ_MAIN_FUNCTION -I. -I./minisketch/include -I./secp256k1/include -I./univalue/include -I./leveldb/include -isystem /tmp/cirrus-ci-build/depends/arm-linux-gnueabihf/include -DBOOST_MULTI_INDEX_DISABLE_SERIALIZATION -DBOOST_NO_CXX98_FUNCTION_BASE   -isystem /tmp/cirrus-ci-build/depends/arm-linux-gnueabihf/include -pthread -I/tmp/cirrus-ci-build/depends/arm-linux-gnueabihf/include  -I/tmp/cirrus-ci-build/depends/arm-linux-gnueabihf/include/  -fdebug-prefix-map=/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-arm-linux-gnueabihf=. -fstack-reuse=none -Wstack-protector -fstack-protector-all -fstack-clash-protection   -Werror    -fno-extended-identifiers -fvisibility=hidden -fPIE -pipe -std=c++17 -O2 -Wno-psabi -c -o libbitcoin_node_a-init.o `test -f 'init.cpp' || echo './'`init.cpp
init.cpp: In function ‘bool LockDataDirectory(bool)’:
init.cpp:1021:1: error: control reaches end of non-void function [-Werror=return-type]
 1021 | }
      | ^
cc1plus: all warnings being treated as errors
make[2]: *** [Makefile:10093: libbitcoin_node_a-init.o] Error 1
```

**Comment by TheCharlatan** [comment]
Nice, Concept ACK.

File: src/init.cpp [hunk 1]
```diff
@@ -1012,10 +1012,9 @@ static bool LockDataDirectory(bool probeOnly)
 {
     // Make sure only a single Bitcoin process is using the data directory.
     const fs::path& datadir = gArgs.GetDataDirNet();
-    if (!DirIsWritable(datadir)) {
-        return InitError(strprintf(_("Cannot write to data directory '%s'; check permissions."), fs::PathToString(datadir)));
-    }
     switch (util::LockDirectory(datadir, ".lock", probeOnly)) {
+    case util::LockResult::ErrorWrite:
+        return InitError(strprintf(_("Cannot write to data directory '%s'; check permissions."), fs::PathToString(datadir)));
```
Comment (TheCharlatan) [inline src/init.cpp:1017]: While testing I noticed that the settings file in the datadir is already interacted with before we do this check. So I think the only way to trigger this is if the lock file has bad permissions.

File: src/wallet/bdb.cpp [hunk 2]
```diff
@@ -149,7 +149,7 @@ bool BerkeleyEnvironment::Open(bilingual_str& err)
 
     fs::path pathIn = fs::PathFromString(strPath);
     TryCreateDirectories(pathIn);
-    if (!LockDirectory(pathIn, ".walletlock")) {
+    if (util::LockDirectory(pathIn, ".walletlock") != util::LockResult::Success) {
```
Comment (TheCharlatan) [inline src/wallet/bdb.cpp:152]: Could probably improve the logging a bit here now that there is a diagnostic difference between a locking error and a lock write error.

**Review by sedited — APPROVED** [review]
Nice, ACK fa5bcb6dafe376f81a669bf9324eedecf3feeb8b

**Comment by hebasto** [comment]
Concept ACK.

File: src/test/util_tests.cpp [hunk 3]
```diff
@@ -1122,6 +1122,7 @@ static constexpr char ExitCommand = 'X';
             ch = [&] {
                 switch (util::LockDirectory(dirname, lockname)) {
                 case util::LockResult::Success: return 1;
+                case util::LockResult::ErrorWrite: return 2;
```
Comment (theuni) [inline src/test/util_tests.cpp:1124]: Can you help me understand what 2 means here?

**Review by theuni — COMMENTED** [review]
Concept ACK.

**Review by sedited — APPROVED** [review]
Re-ACK fa3da629a1aebcb4500803d7417feed8e34285b0

**Review by hebasto — APPROVED** [review]
ACK fa3da629a1aebcb4500803d7417feed8e34285b0, I have reviewed the code and it looks OK.

**Pull request merged**